src/index/src/inverted_index/create/sort/intermediate_rw.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod codec_v1;
+
+use std::collections::BTreeMap;
+
+use common_base::BitVec;
+use futures::{stream, AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt, StreamExt};
+use futures_codec::{FramedRead, FramedWrite};
+use snafu::ResultExt;
+
+use crate::inverted_index::create::sort::SortedStream;
+use crate::inverted_index::error::{
+    CloseSnafu, FlushSnafu, ReadSnafu, Result, UnknownIntermediateCodecMagicSnafu, WriteSnafu,
+};
+use crate::inverted_index::Bytes;
+
+/// `IntermediateWriter` serializes and writes intermediate data to the wrapped `writer`
+pub struct IntermediateWriter<W> {
+    writer: W,
+}
+
+impl<W: AsyncWrite + Unpin> IntermediateWriter<W> {
+    /// Creates a new `IntermediateWriter` wrapping an `AsyncWrite`
+    pub fn new(writer: W) -> IntermediateWriter<W> {
+        IntermediateWriter { writer }
+    }
+
+    /// Serializes and writes all provided values to the wrapped writer
+    pub async fn write_all(mut self, values: BTreeMap<Bytes, BitVec>) -> Result<()> {
+        let (codec_magic, encoder) = (codec_v1::CODEC_V1_MAGIC, codec_v1::IntermediateCodecV1);
+
+        self.writer
+            .write_all(codec_magic)
+            .await
+            .context(WriteSnafu)?;
+
+        let value_stream = stream::iter(values.into_iter().map(Ok));
+        let frame_write = FramedWrite::new(&mut self.writer, encoder);
+        value_stream.forward(frame_write).await?;
+
+        self.writer.flush().await.context(FlushSnafu)?;
+        self.writer.close().await.context(CloseSnafu)
```

Comment:
If `flush` fails, do we need also call `close`?

Reply:
I think the `Drop` impl of the most underlying writer should do the `close` in such a situation.